File: annotation-processor/src/main/java/io/quarkiverse/jimmer/graphql/apt/JimmerGraphQLSourceModel.java
package io.quarkiverse.jimmer.graphql.apt;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;

final class JimmerGraphQLSourceModel {

    private static final String GRAPHQL_PACKAGE_SUFFIX = ".graphql";

    private final Map<String, JimmerGraphQLSourceType> typesByQualifiedName;

    JimmerGraphQLSourceModel(List<JimmerGraphQLSourceType> types) {
        this.typesByQualifiedName = new LinkedHashMap<>();
        for (JimmerGraphQLSourceType type : types) {
            JimmerGraphQLSourceType previous = typesByQualifiedName.putIfAbsent(type.qualifiedName(), type);
            if (previous != null && !previous.equals(type)) {
                throw new IllegalStateException("Duplicate GraphQL source type: " + type.qualifiedName());
            }
        }
    }

    Collection<JimmerGraphQLSourceType> entities() {
        List<JimmerGraphQLSourceType> entities = new ArrayList<>();
        for (JimmerGraphQLSourceType type : typesByQualifiedName.values()) {
            if (type.isEntity()) {
                entities.add(type);
            }
        }
        return entities;
    }

    JimmerGraphQLSourceType type(String qualifiedName) {
        return typesByQualifiedName.get(qualifiedName);
    }

    boolean isEntityType(String qualifiedName) {
        JimmerGraphQLSourceType type = typesByQualifiedName.get(qualifiedName);
        return type != null && type.isEntity();
    }

    List<JimmerGraphQLSourceMethod> scalarMethods(JimmerGraphQLSourceType type) {
        LinkedHashMap<String, JimmerGraphQLSourceMethod> methods = new LinkedHashMap<>();
        collectMethods(type, methods, false);
        return new ArrayList<>(methods.values());
    }

    List<JimmerGraphQLSourceMethod> complexMethods(JimmerGraphQLSourceType type) {
        LinkedHashMap<String, JimmerGraphQLSourceMethod> methods = new LinkedHashMap<>();
        collectMethods(type, methods, true);
        return new ArrayList<>(methods.values());
    }

    private void collectMethods(
            JimmerGraphQLSourceType type,
            LinkedHashMap<String, JimmerGraphQLSourceMethod> methods,
            boolean complex) {
        for (String parentName : type.extendsTypes()) {
            JimmerGraphQLSourceType parent = typesByQualifiedName.get(parentName);
            if (parent != null && (parent.isEntity() || parent.isMappedSuperclass())) {
                collectMethods(parent, methods, complex);
            }
        }
        for (JimmerGraphQLSourceMethod method : type.methods()) {
            if (method.complex() == complex) {
                methods.put(method.name(), method);
            }
        }
    }

    String facadeClassName(String qualifiedName) {
        JimmerGraphQLSourceType type = type(qualifiedName);
        if (type == null) {
            throw new IllegalArgumentException("Illegal GraphQL source type: " + qualifiedName);
        }
        return type.simpleName() + "Gql";
    }

    String facadeQualifiedName(String qualifiedName) {
        return graphqlPackageName(qualifiedName) + '.' + facadeClassName(qualifiedName);
    }

    List<String> entityQualifiedNames() {
        LinkedHashSet<String> entityNames = new LinkedHashSet<>();
        for (JimmerGraphQLSourceType entity : entities()) {
            entityNames.add(entity.qualifiedName());
        }
        return new ArrayList<>(entityNames);
    }

    Map<String, List<String>> entityQualifiedNamesByGraphqlPackage() {
        Map<String, List<String>> entityNamesByPackage = new LinkedHashMap<>();
        for (JimmerGraphQLSourceType entity : entities()) {
            entityNamesByPackage
                    .computeIfAbsent(graphqlPackageName(entity.qualifiedName()), ignored -> new ArrayList<>())
                    .add(entity.qualifiedName());
        }
        return entityNamesByPackage;
    }

    String graphqlPackageName(String qualifiedName) {
        JimmerGraphQLSourceType type = type(qualifiedName);
        if (type == null) {
            throw new IllegalArgumentException("Illegal GraphQL source type: " + qualifiedName);
        }
        if (type.packageName().isEmpty()) {
            return "graphql";
        }
        if (type.packageName().endsWith(GRAPHQL_PACKAGE_SUFFIX)) {
            return type.packageName();
        }
        return type.packageName() + GRAPHQL_PACKAGE_SUFFIX;
    }
}

File: ksp-processor/src/main/java/io/quarkiverse/jimmer/graphql/ksp/JimmerGraphQLSourceModel.kt
package io.quarkiverse.jimmer.graphql.ksp

class JimmerGraphQLSourceModel(types: List<JimmerGraphQLSourceType>) {

    companion object {
        private const val GRAPHQL_PACKAGE_SUFFIX = ".graphql"
    }

    private val typesByQualifiedName: Map<String, JimmerGraphQLSourceType> =
        LinkedHashMap<String, JimmerGraphQLSourceType>().apply {
            for (type in types) {
                val previous = putIfAbsent(type.qualifiedName, type)
                require(previous == null || previous == type) {
                    "Duplicate GraphQL source type: ${type.qualifiedName}"
                }
            }
        }

    fun entities(): List<JimmerGraphQLSourceType> =
        typesByQualifiedName.values.filter { it.isEntity() }

    fun type(qualifiedName: String): JimmerGraphQLSourceType? =
        typesByQualifiedName[qualifiedName]

    fun isEntityType(qualifiedName: String): Boolean =
        typesByQualifiedName[qualifiedName]?.isEntity() == true

    fun scalarMethods(type: JimmerGraphQLSourceType): List<JimmerGraphQLSourceMethod> =
        collectMethods(type, complex = false)

    fun complexMethods(type: JimmerGraphQLSourceType): List<JimmerGraphQLSourceMethod> =
        collectMethods(type, complex = true)

    fun facadeClassName(qualifiedName: String): String =
        requireNotNull(type(qualifiedName)) { "Illegal GraphQL source type: $qualifiedName" }.simpleName + "Gql"

    fun entityQualifiedNamesByGraphqlPackage(): Map<String, List<String>> =
        linkedMapOf<String, MutableList<String>>().also { entityNamesByPackage ->
            for (entity in entities()) {
                entityNamesByPackage.computeIfAbsent(graphqlPackageName(entity.qualifiedName)) { mutableListOf() }
                    .add(entity.qualifiedName)
            }
        }

    fun graphqlPackageName(qualifiedName: String): String {
        val type = requireNotNull(type(qualifiedName)) { "Illegal GraphQL source type: $qualifiedName" }
        return when {
            type.packageName.isEmpty() -> "graphql"
            type.packageName.endsWith(GRAPHQL_PACKAGE_SUFFIX) -> type.packageName
            else -> type.packageName + GRAPHQL_PACKAGE_SUFFIX
        }
    }

    private fun collectMethods(
        type: JimmerGraphQLSourceType,
        complex: Boolean,
    ): List<JimmerGraphQLSourceMethod> {
        val methods = linkedMapOf<String, JimmerGraphQLSourceMethod>()
        collectMethods(type, methods, complex)
        return methods.values.toList()
    }

    private fun collectMethods(
        type: JimmerGraphQLSourceType,
        methods: LinkedHashMap<String, JimmerGraphQLSourceMethod>,
        complex: Boolean,
    ) {
        for (parentName in type.extendsTypes) {
            val parent = typesByQualifiedName[parentName]
            if (parent != null && (parent.isEntity() || parent.isMappedSuperclass())) {
                collectMethods(parent, methods, complex)
            }
        }
        for (method in type.methods) {
            if (method.complex == complex) {
                methods[method.name] = method
            }
        }
    }
}
